drivers/kscan/kscan_npcx.c

```diff
@@ -0,0 +1,514 @@
+		for (int r = 0; r < CONFIG_KSCAN_NPCX_ROW_SIZE; r++) {
+			uint8_t mask = BIT(r);
+			uint8_t row_bit = matrix_new_state[c] & mask;
+
+			/* Continue if we already debounce a key */
+			if (!(deb_col & mask)) {
+				continue;
+			}
+
+			/* Convert the clock cycle differences to usec */
+			uint32_t debt = k_cyc_to_us_floor32(
+				cycles_now - data->scan_clk_cycle[data->scan_cycle_idx[c][r]]);
+
+			/* Does the key requires more time to be debounced? */
+			if (debt < (row_bit ? data->deb_time_press : data->deb_time_rel)) {
+				/* Need more time to debounce */
+				continue;
+			}
+
+			data->matrix_unstable_state[c] &= ~row_bit;
+
+			/* Check if there was a change in the stable state */
+			if ((data->matrix_stable_state[c] & mask) == row_bit) {
+				/* Key state did not change */
+				continue;
+			}
+
+			/*
+			 * The current row has been debounced, therefore update
+			 * the stable state. Then, proceed to notify the
+			 * application about the keys pressed.
+			 */
+			data->matrix_stable_state[c] ^= mask;
+			if (data->callback) {
+				data->callback(dev, r, c, row_bit ? true : false);
+			}
+		}
+	}
+
+	return key_pressed;
+}
+
+static void kscan_matrix_polling_thread(const struct device *dev, void *dummy2, void *dummy3)
+{
+	struct kscan_npcx_data *const data = dev->data;
+	uint32_t current_cycles;
+	uint32_t cycles_diff;
+	uint32_t wait_period;
+
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	while (true) {
+		/* Enable interrupt of KSI pins */
+		kscan_npcx_resume_detection(dev, true);
+
+		kscan_npcx_drive_column(dev, KEYBOARD_COLUMN_DRIVE_ALL);
+		k_sem_take(&data->poll_lock, K_FOREVER);
+		LOG_DBG("Start KB scan!!");
+
+		/* Disable interrupt of KSI pins and start polling */
+		kscan_npcx_resume_detection(dev, false);
+
+		uint64_t poll_time_end = sys_clock_timeout_end_calc(K_USEC(data->poll_timeout));
+
+		while (atomic_get(&data->enable_scan) == 1U) {
+			uint32_t start_period_cycles = k_cycle_get_32();
+
+			if (check_key_events(dev)) {
+				poll_time_end =
+					sys_clock_timeout_end_calc(K_USEC(data->poll_timeout));
+
+			} else if (k_uptime_ticks() > poll_time_end) {
+				break;
+			}
+
+			/*
+			 * Subtract the time invested from the sleep period
+			 * in order to compensate for the time invested
+			 * in debouncing a key
+			 */
+			current_cycles = k_cycle_get_32();
+			cycles_diff = current_cycles - start_period_cycles;
+			wait_period = data->poll_period - k_cyc_to_us_floor32(cycles_diff);
+
+			/* Override wait_period in case it is less than 1 ms */
+			if (wait_period < USEC_PER_MSEC) {
+				wait_period = USEC_PER_MSEC;
+			}
+
+			/*
+			 * wait period results in a larger number when
+			 * current cycles counter wrap. In this case, the
+			 * whole poll period is used
+			 */
+			if (wait_period > data->poll_period) {
+				LOG_DBG("wait_period : %u", wait_period);
+
+				wait_period = data->poll_period;
+			}
+
+			/* Allow other threads to run while we sleep */
+			k_usleep(wait_period);
```

Comment:
How about moving this block in its own function? Triple indent is harsh, would make it a bit more compact, make that break on line 339 less ambiguous (two nested whie...).

Reply:
Done.